File: game/scripts/vscripts/pve/abilities/units/dire_tower/dire_tower_attack/dire_tower_attack.lua
dire_tower_attack = class({})
LinkLuaModifier("modifier_dire_tower_attack", "pve/abilities/units/dire_tower/dire_tower_attack/modifier_dire_tower_attack", LUA_MODIFIER_MOTION_NONE)

function dire_tower_attack:GetCastAnimationCustom()		return ACT_DOTA_CUSTOM_TOWER_ATTACK end
function dire_tower_attack:GetPlaybackRateOverride()       return 1.0 end
function dire_tower_attack:GetCastPointSpeed() 			return 0 end
function dire_tower_attack:GetCastPoint()
	if IsServer() then
		return self:GetCaster():GetAttackAnimationPoint()
	end
end

function dire_tower_attack:GetIntrinsicModifierName()
    return "modifier_dire_tower_attack"
end

function dire_tower_attack:GetCooldown(iLevel)
	if IsServer() then
        local attacks_per_second = self:GetCaster():GetAttacksPerSecond()
        local attack_speed = (1 / attacks_per_second)
		
		return self.BaseClass.GetCooldown(self, self:GetLevel()) + attack_speed
	end
end

function dire_tower_attack:OnSpellStart()
    local caster = self:GetCaster()
    local point = self:GetCursorPosition()
    local origin = caster:GetOrigin()
    local projectile_direction = Direction2D(origin, point)

    CustomEntitiesLegacy:ProjectileAttack(caster, {
		bIsBasicAttack = true,
        tProjectile = {
            EffectName = "particles/dire_tower/ti10_dire_tower_attack.vpcf",
            vSpawnOrigin = origin + Vector(0, 0, 200),
            fDistance = self:GetCastRange(Vector(0,0,0), nil),
            fStartRadius = 70,
            Source = caster,
            vVelocity = projectile_direction * 1000,
            UnitBehavior = PROJECTILES_DESTROY,
            TreeBehavior = PROJECTILES_NOTHING,
            WallBehavior = PROJECTILES_DESTROY,
            GroundBehavior = PROJECTILES_NOTHING,
            fGroundOffset = 0,
            UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not CustomEntitiesLegacy:Allies(_self.Source, unit) end,
            OnUnitHit = function(_self, unit) 
				CustomEntitiesLegacy:AttackWithBaseDamage(caster, {
					hTarget = unit,
					hAbility = self,
				})
            end,
            OnFinish = function(_self, pos)
                EFX("particles/econ/world/towers/ti10_dire_tower/ti10_dire_tower_attack_impact.vpcf", PATTACH_WORLDORIGIN, nil, {
                    cp0 = pos,
                    cp3 = pos,
                })
                EmitSoundOn('Tower.HeroImpact', caster)
            end,
        }
    })

    EmitSoundOn('Tower.HeroImpact', caster)
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(dire_tower_attack)

File: src/vscripts/pve/abilities/units/dire_tower_attack.ts
import { CustomAbility } from "../../../abilities/framework/custom_ability";
import { CustomModifier } from "../../../abilities/framework/custom_modifier";
import { registerAbility, registerModifier } from "../../../lib/dota_ts_adapter";
import { attackWithBaseDamage, direction2D } from "../../../util";

@registerAbility("dire_tower_attack")
class DireTowerAttack extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_CUSTOM_TOWER_ATTACK;
    }

    GetPlaybackRateOverride() {
        return 1.0;
    }

    GetCastingCrawl() {
        return 0;
    }

    GetCastPoint() {
        if (IsServer()) {
            return this.caster.GetAttackAnimationPoint();
        }
        return 0;
    }

    GetIntrinsicModifierName() {
        return ModifierDireTowerAttack.name;
    }

    GetCooldown(level: number) {
        if (IsServer()) {
            const attacksPerSecond = this.caster.GetAttacksPerSecond();
            const attackSpeed = 1 / attacksPerSecond;

            return super.GetCooldown(level) + attackSpeed;
        }

        return super.GetCooldown(level);
    }

    OnSpellStart() {
        const point = this.GetCursorPosition();
        const origin = this.caster.GetOrigin();
        const projectileDirection = direction2D(origin, point);
        const projectileSpeed = 1000;

        this.ProjectileAttack({
            source: this.caster,
            attackType: "basic",
            effectName: "particles/dire_tower/ti10_dire_tower_attack.vpcf",
            distance: this.GetCastRange(Vector(0, 0, 0), undefined),
            startRadius: 70,
            spawnOrigin: origin.__add(Vector(0, 0, 200)),
            velocity: projectileDirection.__mul(projectileSpeed),
            groundOffset: 0,
            unitTest: (unit, projectile) =>
                unit.GetUnitName() != "npc_dummy_unit" && !CustomEntitiesLegacy.Allies(projectile.getSource(), unit),
            onUnitHit: (unit, projectile) => {
                attackWithBaseDamage({
                    source: projectile.getSource(),
                    target: unit,
                    ability: this
                });
            },
            onFinish: projectile => {
                EFX(
                    "particles/econ/world/towers/ti10_dire_tower/ti10_dire_tower_attack_impact.vpcf",
                    ParticleAttachment.WORLDORIGIN,
                    undefined,
                    {
                        cp0: projectile.getPosition(),
                        cp3: projectile.getPosition()
                    }
                );
                EmitSoundOn("Tower.HeroImpact", this.caster);
            }
        });

        EmitSoundOn("General.Attack", this.caster);
    }
}

@registerModifier({ customNameForI18n: "modifier_dire_tower_attack" })
class ModifierDireTowerAttack extends CustomModifier {
    OnCreated() {
        if (IsServer()) {
            EFX(
                "particles/econ/world/towers/ti10_dire_tower/ti10_dire_tower_ambient.vpcf",
                ParticleAttachment.ABSORIGIN_FOLLOW,
                this.parent,
                {
                    release: true
                }
            );
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.ON_DEATH];
    }

    OnDeath(event: ModifierInstanceEvent) {
        if (IsServer() && event.unit == this.parent) {
            this.parent.StartGestureWithPlaybackRate(GameActivity.DOTA_CUSTOM_TOWER_DIE, 0.5);
        }
    }
}
